File: src/main/java/io/github/bot/terminal/application/domains/monitoring/repository/MonitoringTaskRepository.java
package io.github.bot.terminal.application.domains.monitoring.repository;


import lombok.RequiredArgsConstructor;
import org.dizitart.no2.objects.ObjectRepository;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;

import static org.dizitart.no2.objects.filters.ObjectFilters.eq;

@Repository
@RequiredArgsConstructor
public class MonitoringTaskRepository {

    private final ObjectRepository<MonitoringTaskDetails> db;
    private final ConcurrentMap<String, ReentrantLock> locks = new ConcurrentHashMap<>();

    public void add(MonitoringTaskDetails details) {
        details.setId(UUID.randomUUID().toString());
        db.insert(details);
    }

    public void update(MonitoringTaskDetails details) {
        db.update(details);
    }

    public Optional<MonitoringTaskDetails> findById(String id) {
        return Optional.ofNullable(db.find(eq("id", id)).firstOrDefault());
    }

    public List<MonitoringTaskDetails> findAll() {
        return db.find().toList();
    }

    public void deleteById(String id) {
        transaction(id, (repository) -> {
            if (repository.findById(id).isPresent()) {
                db.remove(eq("id", id));
            }
        });
    }

    public void transaction(String id, Consumer<MonitoringTaskRepository> action) {
        Lock lock = getLock(id);
        lock.lock();
        try {
            action.accept(this);
        } finally {
            lock.unlock();
        }
    }

    private Lock getLock(String id) {
        return locks.computeIfAbsent(id, (k) -> new ReentrantLock());
    }
}

File: src/main/kotlin/io/github/bot/terminal/application/domains/monitoring/repository/MonitoringTaskRepository.kt
package io.github.bot.terminal.application.domains.monitoring.repository

import org.dizitart.no2.objects.ObjectRepository
import org.dizitart.no2.objects.filters.ObjectFilters
import org.springframework.stereotype.Service
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

@Service
class MonitoringTaskRepository(
        val db: ObjectRepository<MonitoringTaskDetails>
) {
    private val locks: ConcurrentMap<String, ReentrantLock> = ConcurrentHashMap()

    fun add(details: MonitoringTaskDetails) {
        db.insert(details)
    }

    fun update(details: MonitoringTaskDetails) {
        db.update(details)
    }

    fun findById(id: String): MonitoringTaskDetails? {
        return db.find(ObjectFilters.eq("id", id)).firstOrNull()
    }

    fun findAll(): List<MonitoringTaskDetails> {
        return db.find().toList()
    }

    fun deleteById(id: String) {
        transaction(id) { repository: MonitoringTaskRepository ->
            repository.findById(id)?.let {
                db.remove(ObjectFilters.eq("id", id))
            }
        }
    }

    fun transaction(id: String, action: (MonitoringTaskRepository) -> Unit) {
        val lock = getLock(id)
        lock.lock()
        try {
            action(this)
        } finally {
            lock.unlock()
        }
    }

    private fun getLock(id: String): Lock {
        return locks.computeIfAbsent(id) { k: String -> ReentrantLock() }
    }
}